File: epi_judge_python/k_largest_values_in_bst.py
from typing import List

from bst_node import BstNode
from test_framework import generic_test, test_utils


def find_k_largest_in_bst(tree: BstNode, k: int) -> List[int]:
    # TODO - you fill in here.

    # just do a backwards inorder traversal
    nodeOfInterest = None
    k = k
    result = []

    def decreasingBST(tree):
        if tree is None:
            return
        
        nonlocal k
        if k == 0:
            return

        decreasingBST(tree.right)
        k -= 1  # at max val we do largest, then second largest
        if k >= 0:
            result.append(tree.data)
        else:
            return

        decreasingBST(tree.left)

    decreasingBST(tree)

    return result


if __name__ == '__main__':
    exit(
        generic_test.generic_test_main('k_largest_values_in_bst.py',
                                       'k_largest_values_in_bst.tsv',
                                       find_k_largest_in_bst,
                                       test_utils.unordered_compare))

File: epi_judge_cpp/k_largest_values_in_bst.cc
#include <memory>
#include <vector>

#include "bst_node.h"
#include "test_framework/generic_test.h"
using std::unique_ptr;
using std::vector;

void FindKLargestInBSTHelper(const unique_ptr<BstNode<int>>& tree,int k,vector<int>& v){

  if (tree==nullptr){
    return;
  }
  
  FindKLargestInBSTHelper(tree->right,k,v);
  if (v.size()==k){
    // before adding anything else to our result list
    // check that we have not already found k items
    // if we have then just terminate the search
    return;
  }
  v.push_back(tree->data);
    
  FindKLargestInBSTHelper(tree->left,k,v);
  
  
}

vector<int> FindKLargestInBST(const unique_ptr<BstNode<int>>& tree, int k) {
  // TODO - you fill in here.
  vector<int> v;
  FindKLargestInBSTHelper(tree,k,v);
  return v;
}

int main(int argc, char* argv[]) {
  std::vector<std::string> args{argv + 1, argv + argc};
  std::vector<std::string> param_names{"tree", "k"};
  return GenericTestMain(args, "k_largest_values_in_bst.cc",
                         "k_largest_values_in_bst.tsv", &FindKLargestInBST,
                         UnorderedComparator{}, param_names);
}
